core/modules/replica/HttpIngestModule.cc

```diff
@@ -86,6 +86,43 @@ string leastLoadedWorker(DatabaseServices::Ptr const& databaseServices,
     return worker;
 }
 
+/**
+ * The optimized version of the previously defined function will
+ * re-use and update the transient replica disposition cache when
+ * selecting a candidate worker. Each worker's entry in the cache
+ * will be populated from the database first time 
+ *
+ * @param worker2replicasCache  the cache of replica disposition for workers
+ * @param databaseServices  the service used for pulling the initial replica
+ *   disposition counters for workers.
+ * @param workers  a collection of workers to be evaluated as candidates.
+ * @return the name of a worker which has the least number of replicas
+ */
+template<typename COLLECTION_OF_WORKERS>
+string leastLoadedWorker(map<string,size_t>& worker2replicasCache,
+                         DatabaseServices::Ptr const& databaseServices,
+                         COLLECTION_OF_WORKERS const& workers) {
+    string worker;
+    string const noSpecificDatabase;
+    bool   const allDatabases = true;
+    size_t numReplicas = numeric_limits<size_t>::max();
+    for (auto&& candidateWorker: workers) {
+        if (worker2replicasCache.count(candidateWorker) == 0) {
+            worker2replicasCache[candidateWorker] =
```

Comment:
I don't know how frequently this is called or for how many `workers`, but `::insert` in this case would be slightly more efficient.

Reply:
You have convinced me 👍 I have reimplemented this function in terms of the iterators. Though, after testing the new new code I see no difference in its performance as compared with the `[]`-based implementation. As I already mentioned in some previous comment, taking away expensive MySQL queries against a table of millions of replicas (`0.1s` per query) was the main goal of this ticket. This optimization alone has allowed to speed up the chunk placement code by a factor of x100. Replacing `[]` with iterators made no measurable improvement to the performance. Though, it probably made a sense from the code stye point of view in order to make everything ore "C++-ish".